File: src/utils/style.ts
import { inObject } from './object';
import { Beer } from '../types';

const barrelAged = (beer: Beer): string => (beer.barrelAged ? 'Barrel Aged' : '');

const brewedIn = (beer: Beer): string =>
  inObject('barrelType', beer) ? `in ${beer.barrelType} Barrels` : '';

const brewedWith = (beer: Beer): string =>
  inObject('brewedWith', beer) ? `with ${beer.brewedWith}` : '';

export const formatStyle = (beer: Beer): string =>
  `${barrelAged(beer)} ${beer.style} ${brewedWith(beer)} ${brewedIn(beer)}`;

File: src/formatting/style.rs
use crate::storage::Beer;
use implicit_clone::unsync::IString;

fn barrel_aged(beer: &Beer) -> IString {
    if beer.barrel_aged {
        IString::from("Barrel Aged")
    } else {
        IString::default()
    }
}

fn brewed_in(beer: &Beer) -> IString {
    match &beer.barrel_type {
        Some(barrel_type) => IString::from(format!("in {barrel_type} Barrels")),
        None => IString::default(),
    }
}

fn brewed_with(beer: &Beer) -> IString {
    match &beer.brewed_with {
        Some(brewed_with) => IString::from(format!("with {brewed_with}")),
        None => IString::default(),
    }
}

pub fn format(beer: &Beer) -> IString {
    IString::from(
        format!(
            "{} {} {} {}",
            barrel_aged(beer),
            beer.style,
            brewed_with(beer),
            brewed_in(beer)
        )
        .trim()
        .replace("  ", " "),
    )
}
